plugins/inputs/openstack/openstack.go

```diff
@@ -0,0 +1,1059 @@
+	//if !o.services.ContainsService(volumeV2Service) {
+	//	return nil
+	//}
+	page, err := volumes.List(o.volume, &volumes.ListOpts{AllTenants: true}).AllPages()
+	if err != nil {
+		return fmt.Errorf("unable to list volumes: %v", err)
+	}
+	v := []volume{}
+	volumes.ExtractVolumesInto(page, &v)
+	for _, volume := range v {
+		o.volumes[volume.ID] = volume
+	}
+	return nil
+}
+
+// gatherStoragePools collects storage pools from the OpenStack API.
+func (o *OpenStack) gatherStoragePools() error {
+	//if !o.services.ContainsService(volumeV2Service) {
+	//	return nil
+	//}
+	results, err := schedulerstats.List(o.volume, &schedulerstats.ListOpts{Detail: true}).AllPages()
+	if err != nil {
+		return fmt.Errorf("unable to list storage pools: %v", err)
+	}
+	storagePools, err := schedulerstats.ExtractStoragePools(results)
+	if err != nil {
+		return fmt.Errorf("unable to extract storage pools: %v", err)
+	}
+	for _, storagePool := range storagePools {
+		o.storagePools[storagePool.Name] = storagePool
+	}
+	return nil
+}
+
+// accumulateServerDiagnostics accumulates statistics from the compute(nova) service.
+func (o *OpenStack) accumulateServerDiagnostics(acc telegraf.Accumulator) {
+	for server_id, diagnostic := range o.diag {
+		tags := tagMap{
+			"server_id": server_id,
+		}
+		fields := fieldMap{
+			"server_id": server_id,
+		}
+		switch reflect.TypeOf(diagnostic).Kind() {
+		case reflect.Map:
+			s := reflect.ValueOf(diagnostic).MapRange()
+			port_map := make(map[string]string)
+			port_count := 0
+			for s.Next() {
+				if strings.Contains(s.Key().Interface().(string), "tap") {
+					tap_list := strings.Split(s.Key().Interface().(string), "_")
+					if val, ok := port_map[tap_list[0]]; ok {
+						fields["port_"+val+"_"+strings.Join(tap_list[1:], "_")] = s.Value().Interface().(float64)
+						tags["port_"+val] = tap_list[0]
+					} else {
+						port_count += 1
+						port_map[tap_list[0]] = strconv.Itoa(port_count)
+						fields["port_"+port_map[tap_list[0]]+"_"+strings.Join(tap_list[1:], "_")] = s.Value().Interface().(float64)
+						tags["port_"+port_map[tap_list[0]]] = tap_list[0]
+					}
+				} else {
+					fields[s.Key().Interface().(string)] = s.Value().Interface().(float64)
+				}
+			}
+			fields["no_of_ports"] = port_count
+		}
+		acc.AddFields("openstack_server_diagnostics", fields, tags)
+	}
+}
+
+// accumulateStacks accumulates statistics from the stack service.
+func (o *OpenStack) accumulateStacks(acc telegraf.Accumulator) {
+	for _, stack := range o.stacks {
+		tags := tagMap{
+			"id": stack.ID,
+		}
+		fields := fieldMap{
+			"creation_time":       stack.CreationTime.Format("2006-01-02T15:04:05.999999999Z07:00"),
+			"description":         stack.Description,
+			"id":                  stack.ID,
+			"stack_name":          stack.Name,
+			"stack_status":        stack.Status,
+			"stack_status_reason": stack.StatusReason,
+			"updated_time":        stack.UpdatedTime.Format("2006-01-02T15:04:05.999999999Z07:00"),
```

Comment:
Please think about which information can be a `tag`. I guess `stack.Name` or `stack.Status` are good candidates without growing the cardinality too much.